intel/intel-extension-for-transformers
# [LLM Runtime] Unify KV_cache and Support Batch-dim Process in Beam Search

## Type of Change

feature or bug fix or documentation or others: Feature
API changed or not: YES (`model_eval`)
add a new struct for batch inference:
```c++
struct model_input {
  // embd or next token
  model_token* tokens = nullptr;
  // tokens length
  uint32_t n_tokens = 0;
  // prompt length
  uint32_t n_prompt_tokens = 0;
  // kv cache n_past
  uint32_t n_past = 0;
  // text tokens length (prompt + all next tokens)
  // the number of tokens evaluated so far (including evicted tokens if there is any)
  uint32_t n_total = 0;
  // request id
  int request_idx = -1;
  // beam id in beam search
  int beam_idx = 0;
  // padding related, attention mask
  // (0: left, 1: right)
  // only support padding left in decoder only model
  int padding_side = 0;
  // padding length
  uint32_t n_padding = 0;
};

MODEL_API int model_eval(struct model_context* ctx, const model_input* inputs, const int n_input, int n_threads););
```
cc @Zhenzhong1 , we can co-work and discuss about how to support static batching generation since you are doing `ChatGLM` multi-batch inference (pad, mask MHA, return, etc.)

## Description

detail description 
JIRA ticket: [829](https://jira.devtools.intel.com/browse/NLPTOOLKIU-829)
TODO: 
  - [x] `kv_cache_seq_cpy` and `kc_cache_seq_concat` prototype and refact beam search
  - [x] beam search for static batching prompt
  - [x] deal with padding
  - [x] pybind and python ut example

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)
`transformers` batch_decode example (**padding_side='left'** for right next_token generation)
```python
from transformers import pipeline, set_seed, AutoModelForCausalLM, AutoTokenizer
model_dir = "/home/zhentao/gpt-j-6b"
tokenizer = AutoTokenizer.from_pretrained(model_dir, padding_side="left")
tokenizer.pad_token = tokenizer.eos_token # to avoid an error
model = AutoModelForCausalLM.from_pretrained(model_dir)
model.eval()

prompts = ["she opened the door and see", 
           "tell me 10 things about jazz music",
           "What is the meaning of life?",
           "To be, or not to be, that is the question: Whether 'tis nobler in the mind to suffer"\
            " The slings and arrows of outrageous fortune, "\
            "Or to take arms against a sea of troubles."\
            "And by opposing end them. To die—to sleep,",]
inputs = tokenizer(prompts, padding=True, return_tensors='pt')
print(inputs)
out = model(**inputs)

generate_ids = model.generate(**inputs, max_new_tokens=128, min_new_tokens=30, early_stopping=True, num_beams=4)
ans = tokenizer.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)
print(ans)

"""
ans[0]
'she opened the door and see me standing there.\n\n"What are you doing here?" she asked.\n\n"I came to see you," I said.\n\n"I don\'t want to see you," she said.\n\n"Why not?" I asked.\n\n"Because I don\'t want to see you," she said.\n\n"Why not?" I asked.\n\n"Because I don\'t want to see you," she said.\n\n"Why not?" I asked.\n\n"Because I don\'t want to see you," she said.\n\n"Why not?" I asked.\n\n"Because I'
ans[1]
'tell me 10 things about jazz music that i don’t know\n\nI’m not a jazz musician. I’m not a jazz fan. I don’t listen to jazz music. I don’t know anything about jazz music. I don’t know anything about jazz musicians. I don’t know anything about jazz fans. I don’t know anything about jazz. I don’t know anything about jazz musicians. I don’t know anything about jazz fans. I don’t know anything about jazz. I don’t know anything about jazz musicians. I don’t'
ans[2]
'What is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\nWhat is the meaning of life?\n\n'
ans[3]
"To be, or not to be, that is the question: Whether 'tis nobler in the mind to suffer The slings and arrows of outrageous fortune, Or to take arms against a sea of troubles.And by opposing end them. To die—to sleep, No more; and by a sleep to say we end The heart-ache and the thousand natural shocks That flesh is heir to—'tis a consummation Devoutly to be wish'd. To die, to sleep; To sleep: perchance to dream: ay, there's the rub; For in that sleep of death what dreams may come, When we have shuffled off this mortal coil, Must give us pause: there's the respect That makes calamity of so long life; For who would bear the whips and scorns of time, The oppressor's wrong, the proud man's contumely, The"
"""
```

## Dependency Change?

any library dependency introduced or removed
